File: 04-estructuras-datos/04-pilas/scala/implementaciones/PilaEnlazada.scala
final class PilaEnlazada[A]:
  private final class Nodo(val dato: A, val siguiente: Nodo | Null)
  private var cima: Nodo | Null = null

  def push(dato: A): Unit = cima = new Nodo(dato, cima)

  def pop(): A =
    if cima == null then throw new NoSuchElementException("Pila vacía")
    val actual = cima.asInstanceOf[Nodo]
    // La cima avanza al siguiente nodo después de retirar el actual.
    cima = actual.siguiente
    actual.dato

File: 04-estructuras-datos/04-pilas/java/implementaciones/PilaEnlazada.java
import java.util.NoSuchElementException;

public class PilaEnlazada<T> {
    private static class Nodo<T> {
        T dato;
        Nodo<T> siguiente;
        Nodo(T dato, Nodo<T> siguiente) { this.dato = dato; this.siguiente = siguiente; }
    }

    private Nodo<T> cima;

    public void push(T dato) {
        // El nuevo nodo apunta a la antigua cima y pasa a ser la nueva cima.
        cima = new Nodo<>(dato, cima);
    }

    public T pop() {
        if (cima == null) throw new NoSuchElementException("Pila vacía");
        T dato = cima.dato;
        cima = cima.siguiente;
        return dato;
    }

    public T peek() {
        if (cima == null) throw new NoSuchElementException("Pila vacía");
        return cima.dato;
    }
}
